### PR TITLE
OpenSSL::Test: bring back the relative paths

### LINKED ISSUE #10628: Test.pm: regression caused by the Perl abs2rel workaround
While working on https://github.com/openssl/openssl/issues/10604#issuecomment-564458060, I noticed that the tests now behave differently from what I was accustomed to: all paths in the test commands are given as absolute paths:

```
~/src/openssl$ make V=1 TESTS=test_ocsp tests
...
/home/msp/src/openssl/util/shlib_wrap.sh /home/msp/src/openssl/apps/openssl base64 -d -in /home/msp/src/openssl/test/ocsp-tests/ND1.ors -out ocsp-resp-fff.dat => 0
```

Until recently, the paths used to be relative to the test directory:

```
~/src/openssl$ make V=1 TESTS=test_ocsp tests
...
../../util/shlib_wrap.sh ../../apps/openssl base64 -d -in ../../test/ocsp-tests/ND1.ors -out ocsp-resp-fff.dat => 0
```

This was very handy for several reasons: For example, it was easy to copy & paste a command from a [verbose test run](https://github.com/openssl/openssl/issues/10604#issuecomment-564458060) of someone else into your shell (or from your master worktree into your 1.1.1. worktree) and it could be **run without editing**. If necessary, the command could also be debugged by inserting in `gdb --args` in the right location:

```
~/src/openssl$ make V=1 TESTS=test_ocsp tests
[...]
../../util/shlib_wrap.sh gdb --args ../../apps/openssl base64 -d -in ../../test/ocsp-tests/ND1.ors -out ocsp-resp-fff.dat => 0
```

The cause for this behavioral change is the following commit which added a workaraund for a Perl abs2rel bug:

* 7a2027240e1d01f7f5b209998d1de36af221b34b Adding Test.pm with workaround for Perl abs2rel bug [master]
* 374cab6390ba005d4a559a3dea2a034af9cb1c09 Adding Test.pm with workaround for Perl abs2rel bug [1.1.1]

I consider this behavioral change a regression, because it impedes the process of analyzing test failures. That's why I tagged it as a bug and I would appreciate very much if @mpeylo and @levitte would find an alternative workaround which does not have this negative side effect.

### PR DESCRIPTION
**Pull request by levitte**
Because there was a bug in File::Spec::Unix' abs2rel when it was given
relative paths as both PATH and BASE arguments, the directories we
deal with were made to be all absolute.  Unfortunately, this meant
getting paths in our verbose test output which are difficult to use
anywhere else (such as a separate test build made for comparison), due
to the constant need to edit all the paths all the time.

We're therefore getting back the relative paths, by doing an extra
abs2rel() in __srctop_file, __srctop_dir, __bldtop_file and
__bldtop_dir, with a 'Cwd::getcwd' call as BASE argument.

Fixes #10628

### CODE REVIEW

**Review by mattcaswell — APPROVED** [review]

**Review by mspncp — CHANGES_REQUESTED** [review]
Please don't merge this one yet...

**Comment by mspncp** [comment]
I would like to do some additional investigations to verify that this pull doesn't reintroduce the bug which #7031 fixed. Therefore I first reverted commit 7a2027240e1d01f7f5b209998d1de36af221b34b and tried to reproduce the bug.

But currently I don't see any bug in the testsuite at all, all I see is a Perl bug. So I'm beginning to wonder whether #7031 was necessary at all, or only a precaution. If that's the case, I would prefer if #7031 would be reverted instead of introducing a fix for an unnecessary fix. But I need some more time to workout the details. Currently I feel a little bit ill, so please be patient, I might not be able to say more until the weekend.

**Comment by t8m** [comment]
@mspncp I think OTC hold label should have been used instead of removing the ready to merge label. Putting the label there does not mean you need to start a vote right ahead.

**Comment by mspncp** [comment]
You are right, thanks for the hint. In fact, I was undecided which option to choose.

**Comment by levitte** [comment]
To be noted is that the Perl bug still exist.  I verified that just now...  and this PR doesn't fix it, as far as I can tell (my assumption that it was only affected when relative paths were the input was wrong)

**Comment by levitte** [comment]
Verification of the bug:

``` console
$ perl -e 'use File::Spec qw(abs2rel); print File::Spec->abs2rel("/home/goal/test/..", "/home/goal/test/../../base"), "\n";'
../..
```

So, errrr, back to the drawing table

**Comment by levitte** [comment]
What is unquestionable is that this Perl bug affects the performance of our tests in circumstances we have no control over, so we do need to counter it *somehow*.

I'm currently pondering simply adding a replacement abs2rel, as soon as I figured out what the bug is.  Yay for open source :wink:

**Comment by levitte** [comment]
Ah, hold on...  this, *together with* #7031, should actually fix the problem.  The #7031 fix ensures that we get absolute paths with no '..' components anywhere, which means that the `abs2rel` call that this PR adds can calculate the relative path without running into the issue that '..' components can cause.

**Comment by richsalz** [comment]
We don't need to fix the problem globally, and we don't need to do a fix that can go upstream.  We just need to fix it in our tests on unix; the other platforms can continue to use what is currrently check-in

So perhaps a simple "replace stuff/foo/../otherstuff" with "stuff/otherstuff" loop?

**Comment by levitte** [comment]
> So perhaps a simple "replace stuff/foo/../otherstuff" with "stuff/otherstuff" loop?

That's what one might imagine `File::Spec->canonpath()` does...  but it's not quite that simple, to the point that this function refuses to touch '..' components unless they're at the beginning or the end of the path.  Symlinks are not your friends...

**Comment by richsalz** [comment]
Don't we already have issues with symlinks?  Like I said, we're not trying to be the real answer.

**Comment by mspncp** [comment]
I don't question the presence of the _Perl" bug, I just have my doubts whether _OpenSSL_ was affected before 7a2027240e1d01f7f5b209998d1de36af221b34b.

Please take a closer look at [Perl RT #133465][RT133465] (migrated to https://github.com/Perl/perl5/issues/16666), in particular the two replies [here][RT133465-1] and [here][RT133465-2] from the maintainer. According to his analysis, the problem seems to occur only if the target directory has trailing `\..`. So as you already noted, calling `abs2rel()` on pathes which were previously converted to absolute pathes should be harmless. Looking at the source (right before 7a2027240e1d01f7f5b209998d1de36af221b34b), I see many cases of `abs2rel(rel2abs(...),...`:

[RT133465]: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=133465
[RT133465-1]: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=133465#txn-1576702
[RT133465-2]: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=133465#txn-1576980


```make
msp@debian:~/src/openssl$ git grep -n abs2rel 7a2027240e1d01f7f5b209998d1de36af221b34b~
Configurations/descrip.mms.tmpl:5:  use File::Spec::Functions qw/:DEFAULT abs2rel rel2abs/;
Configurations/descrip.mms.tmpl:33:      my $b = abs2rel(rel2abs("."),rel2abs($f));
Configurations/descrip.mms.tmpl:694:  use File::Spec::Functions qw/abs2rel rel2abs catfile catdir/;
Configurations/descrip.mms.tmpl:772:         my $dofile = abs2rel(rel2abs(catfile($config{sourcedir},
Configurations/descrip.mms.tmpl:867:      my $backward = abs2rel(rel2abs("."), rel2abs($forward));
Configurations/descrip.mms.tmpl:868:      my $objd = abs2rel(rel2abs(dirname($obj)), rel2abs($forward));
Configurations/descrip.mms.tmpl:870:      my $depd = abs2rel(rel2abs(dirname($dep)), rel2abs($forward));
Configurations/descrip.mms.tmpl:873:          join(", ", map { abs2rel(rel2abs($_), rel2abs($forward)) } @srcs);
Configurations/descrip.mms.tmpl:971:      my $translatesyms_pl = abs2rel(rel2abs(catfile($config{sourcedir},
Configurations/descrip.mms.tmpl:1170:      my $dofile = abs2rel(rel2abs(catfile($config{sourcedir},
Configurations/unix-Makefile.tmpl:931:  use File::Spec::Functions qw/:DEFAULT abs2rel rel2abs/;
Configurations/unix-Makefile.tmpl:961:              my $dofile = abs2rel(rel2abs(catfile($config{sourcedir},
Configurations/unix-Makefile.tmpl:1228:      my $dofile = abs2rel(rel2abs(catfile($config{sourcedir},
Configurations/windows-makefile.tmpl:521: use File::Spec::Functions qw/:DEFAULT abs2rel rel2abs/;
Configurations/windows-makefile.tmpl:544:          my $mkdef = abs2rel(rel2abs(catfile($config{sourcedir},
Configurations/windows-makefile.tmpl:557:              my $dofile = abs2rel(rel2abs(catfile($config{sourcedir},
Configurations/windows-makefile.tmpl:779:      my $dofile = abs2rel(rel2abs(catfile($config{sourcedir},
Configure:18:use File::Spec::Functions qw/:DEFAULT abs2rel rel2abs splitdir/;
Configure:214:my $dofile = abs2rel(catfile($srcdir, "util/dofile.pl"));
Configure:218:$config{sourcedir} = abs2rel($srcdir);
Configure:219:$config{builddir} = abs2rel($blddir);
Configure:1632:        my $res = abs2rel(absolutedir($dir), rel2abs($relativeto));
Configure:1650:        my $res = abs2rel(catfile(absolutedir($d), $f), rel2abs($relativeto));
Configure:1754:        push @{$config{build_infos}}, catfile(abs2rel($sourced, $blddir), $f);
Configure:1766:                                         builddir => abs2rel($buildd, $blddir),
Configure:1767:                                         sourcedir => abs2rel($sourced, $blddir),
Configure:1768:                                         buildtop => abs2rel($blddir, $blddir),
Configure:1769:                                         sourcetop => abs2rel($srcdir, $blddir) },
```
The only place where I'm not sure is `test/run_tests.pl:87`, where the test harness is called. Maybe adding a `abs2rel(rel2abs($_), rel2abs(curdir()));` there would be sufficient?

```make
test/run_tests.pl:17:use File::Spec::Functions qw/catdir catfile curdir abs2rel rel2abs/;
test/run_tests.pl:87:my $ret = $harness->runtests(map { abs2rel($_, rel2abs(curdir())); }
util/perl/OpenSSL/Test.pm:65:                             catdir catfile splitpath catpath devnull abs2rel
util/perl/OpenSSL/Test.pm:1078:    my $reverse = abs2rel($abscurdir, $absdir);
util/perl/OpenSSL/Test.pm:1111:            my $newpath = abs2rel(rel2abs($directories{$_}), rel2abs($dir));
util/perl/OpenSSL/Test.pm:1121:            my $newpath = abs2rel(rel2abs($ENV{$_}), rel2abs($dir));
```


But my real point is: Before we continue talking about fixes, I would like to know: where is the _OpenSSL  bug_ we are fixing? I.e., I would like to have a reproducible failure. I tried to follow the description in the commit message of 7a2027240e1d01f7f5b209998d1de36af221b34b, but did not succeed to find any misbehaviour (test failing, etc.). While the commit message is very clear about the Perl bug and even gives a test case, it remains vague about how to reproduce the issue in OpenSSL:

```
If SRCTOP != BLDTOP, and SRCTOP is given in relative form, e.g.
"./config ../openssl", then a bug in Perl's abs2rel may trigger that directory-
rewriting in __cwd results in wrong entries in %directories under certain
circumstances, e.g. when a test executes run(app(["openssl"]) after indir.
```

The cited command `./config ../openssl` is wrong, I assume it should be something like `../openssl/config`. So i tried placing the builddir next to the source directory

```
mkdir ~/src/test
cd ~/src/builddir
../openssl/confg
make && make test
```

and even into the source directory

```
mkdir ~/src/openssl/builddir
../config
make && make test
```

but I did not see any misbehaviour of the test suite. Maybe @mpeylo can give us more details about the issue he had?

**Comment by mspncp** [comment]
> Ah, hold on... this, _together with_ #7031, this PR should actually fix the problem. The #7031 fix ensures that we get absolute paths with no '..' components anywhere, which means that the `abs2rel` call that this PR adds can calculate the relative path without running into the issue that '..' components can cause.

As I wrote in my previous reply: maybe we already safe before #7031

> What is unquestionable is that this Perl bug affects the performance of our tests in circumstances we have no control over, so we do need to counter it _somehow_.
> 
> I'm currently pondering simply adding a replacement abs2rel, as soon as I figured out what the bug is. Yay for open source

Great to hear that you have ambitions to fix abs2rel. But if you really intend to do it, please don't reinvent the wheel but [get it fixed upstream](https://github.com/Perl/perl5/issues/16666) instead. Because _that's_ what open source is about. ;-)

**Comment by mspncp** [comment]
A propos perl bug: the maintainer provided a testcase (see below) which you can run yourself

```
$ prove 133465-abs2rel-func.3.t

133465-abs2rel-func.3.t .. 1/? 
#   Failed test 'abs2rel() calculated relative path (../..) as expected (../t2)'
#   at 133465-abs2rel-func.3.t line 262.
#          got: '../..'
#     expected: '../t2'

#   Failed test 'chdir via calculated relative path (../..) brought us to /tmp/ccKLwm8FCX/testing/t1/t2, as expected'
#   at 133465-abs2rel-func.3.t line 266.
#          got: '/tmp/ccKLwm8FCX/testing'
#     expected: '/tmp/ccKLwm8FCX/testing/t1/t2'

#   Failed test 'abs2rel() calculated relative path (../../t2/t3/..) as expected (../../t2)'
#   at 133465-abs2rel-func.3.t line 262.
#          got: '../../t2/t3/..'
#     expected: '../../t2'

#   Failed test 'abs2rel() calculated relative path (../../t2/t3/../../t4/t5) as expected (../../t4/t5)'
#   at 133465-abs2rel-func.3.t line 262.
#          got: '../../t2/t3/../../t4/t5'
#     expected: '../../t4/t5'

#   Failed test 'abs2rel() calculated relative path (../../t2/t3/../../t4/t5/..) as expected (../../t4/t5/..)'
#   at 133465-abs2rel-func.3.t line 262.
#          got: '../../t2/t3/../../t4/t5/..'
#     expected: '../../t4/t5/..'
# Looks like you failed 5 tests of 42.
133465-abs2rel-func.3.t .. Dubious, test returned 5 (wstat 1280, 0x500)
Failed 5/42 subtests 

Test Summary Report
-------------------
133465-abs2rel-func.3.t (Wstat: 1280 Tests: 42 Failed: 5)
  Failed tests:  9-10, 19, 30, 41
  Non-zero exit status: 5
Files=1, Tests=42,  0 wallclock secs ( 0.01 usr  0.04 sys +  0.07 cusr  0.14 csys =  0.26 CPU)
Result: FAIL
```



```perl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Path 2.15 qw(make_path);
use File::Spec;
use File::Temp qw(tempdir);
use Cwd;
use Test::More;
#use Data::Dump qw(dd pp);

my $od = cwd();
{
my $msg = <<END_OF_NOTE;
Case 1: Elaboration of case reported in RT 133465:
('t1/t2/t3/..', 't1/t2/t3/../../t4')
i.e., updir at end of dest path and in middle of base path
END_OF_NOTE

note($msg);

my $tdir = tempdir(CLEANUP => 1);
my $startdir = startdir_setup($tdir);

my $setup_args = {
startdir => $startdir,
dest_needed => [ 't1/t2/t3' ],
base_needed => 't1/t4',
dest_real => 't1/t2',
relpath_components => [ "..", "t2" ],
};
my $expected_relpath = manual_directory_setup($setup_args);

my $test_args = {
startdir => $startdir,
test_dest => 't1/t2/t3/..',
test_base => 't1/t2/t3/../../t4',
};
my @e = confirm_test_paths_created($test_args);

my $chdir_args = {
startdir => $startdir,
expected_relpath => $expected_relpath,
e => [ @e ],
%{$setup_args},
};
my $expected_thisdir = confirm_chdir_via_expected_relpath($chdir_args);

my $use_abs2rel_args = {
startdir => $startdir,
e => [ @e ],
expected_relpath => $expected_relpath,
expected_thisdir => $expected_thisdir,
};
calculate_relpath_via_abs2rel($use_abs2rel_args);

chdir $od or die "Unable to get back to where we started: $!";
}

{
my $msg = <<END_OF_NOTE;
Case 2: updirs are at end of dest path; none in base path:
('t1/t2/t3/..', 't1/t5/t6')
abs2rel() calculates *valid* relative path but not *simplest* relative path
END_OF_NOTE

note($msg);

my $tdir = tempdir(CLEANUP => 1);
my $startdir = startdir_setup($tdir);

my $setup_args = {
startdir => $startdir,
dest_needed => [ 't1/t2/t3' ],
base_needed => 't1/t5/t6',
dest_real => 't1/t2',
relpath_components => [ "..", "..", "t2" ],
};
my $expected_relpath = manual_directory_setup($setup_args);

my $test_args = {
startdir => $startdir,
test_dest => 't1/t2/t3/..',
test_base => 't1/t5/t6',
};
my @e = confirm_test_paths_created($test_args);

my $chdir_args = {
startdir => $startdir,
expected_relpath => $expected_relpath,
e => [ @e ],
%{$setup_args},
};
my $expected_thisdir = confirm_chdir_via_expected_relpath($chdir_args);

my $use_abs2rel_args = {
startdir => $startdir,
e => [ @e ],
expected_relpath => $expected_relpath,
expected_thisdir => $expected_thisdir,
};
calculate_relpath_via_abs2rel($use_abs2rel_args);

chdir $od or die "Unable to get back to where we started: $!";
}

{
my $msg = <<END_OF_NOTE;
Case 3: updirs are in middle of dest path; none in base path:
('t1/t2/t3/../../t4/t5', 't1/t6/t7')
abs2rel() calculates *valid* relative path but not *simplest* relative path
END_OF_NOTE

note($msg);

my $tdir = tempdir(CLEANUP => 1);
my $startdir = startdir_setup($tdir);

my $setup_args = {
startdir => $startdir,
dest_needed => [ 't1/t2/t3', 't1/t4/t5' ],
base_needed => 't1/t6/t7',
dest_real => 't1/t4/t5',
relpath_components => [ "..", "..", "t4", "t5" ],
};
my $expected_relpath = manual_directory_setup($setup_args);

my $test_args = {
startdir => $startdir,
test_dest => 't1/t2/t3/../../t4/t5',
test_base => 't1/t6/t7',
};
my @e = confirm_test_paths_created($test_args);

my $chdir_args = {
startdir => $startdir,
expected_relpath => $expected_relpath,
e => [ @e ],
%{$setup_args},
};
my $expected_thisdir = confirm_chdir_via_expected_relpath($chdir_args);

my $use_abs2rel_args = {
startdir => $startdir,
e => [ @e ],
expected_relpath => $expected_relpath,
expected_thisdir => $expected_thisdir,
};
calculate_relpath_via_abs2rel($use_abs2rel_args);

chdir $od or die "Unable to get back to where we started: $!";
}

{
my $msg = <<END_OF_NOTE;
Case 4: updirs are both in middle of, and at end of dest path; none in base path:
('t1/t2/t3/../../t4/t5/..', 't1/t6/t7')
abs2rel() calculates *valid* relative path but not *simplest* relative path
END_OF_NOTE

note($msg);

my $tdir = tempdir(CLEANUP => 1);
my $startdir = startdir_setup($tdir);

my $setup_args = {
startdir => $startdir,
dest_needed => [ 't1/t2/t3', 't1/t4/t5' ],
base_needed => 't1/t6/t7',
dest_real => 't1/t4',
relpath_components => [ "..", "..", "t4", "t5", ".." ],
};
my $expected_relpath = manual_directory_setup($setup_args);

my $test_args = {
startdir => $startdir,
test_dest => 't1/t2/t3/../../t4/t5/..',
test_base => 't1/t6/t7',
};
my @e = confirm_test_paths_created($test_args);

my $chdir_args = {
startdir => $startdir,
expected_relpath => $expected_relpath,
e => [ @e ],
%{$setup_args},
};
my $expected_thisdir = confirm_chdir_via_expected_relpath($chdir_args);

my $use_abs2rel_args = {
startdir => $startdir,
e => [ @e ],
expected_relpath => $expected_relpath,
expected_thisdir => $expected_thisdir,
};
calculate_relpath_via_abs2rel($use_abs2rel_args);

chdir $od or die "Unable to get back to where we started: $!";
}

done_testing;

#################### SUBROUTINES ####################

sub startdir_setup {
my $tdir = shift;
my $testingdir = File::Spec->catdir($tdir, 'testing');
my @created = make_path($testingdir, { mode => 0711 });
chdir $testingdir or die "Unable to chdir to $testingdir: $!";
my $startdir = cwd();
pass("Starting out from $startdir");
return $startdir;
}

sub manual_directory_setup {
my $args = shift;
my @d = map { File::Spec->catdir($args->{startdir}, File::Spec->splitdir($_)) }
(@{$args->{dest_needed}}, $args->{base_needed});
my @created = make_path(@d, { mode => 0711 });
my $destdir = File::Spec->catdir($args->{startdir}, File::Spec->splitdir($args->{dest_real}));
for my $d (@d, $destdir) {
ok(-d $d, "Created $d for testing");
}
chdir $d[1] or die "Unable to chdir to $d[1]: $!";
my $expected_relpath = File::Spec->catdir(@{$args->{relpath_components}});
chdir $expected_relpath or die "Unable to chdir via relative path: $!";
my $thisdir = cwd();
is($thisdir, $destdir,
"chdir via relative path '$expected_relpath' brought us to $destdir, as expected");
chdir $args->{startdir} or die "Unable to get back to $args->{startdir}: $!";
return $expected_relpath;
}

sub confirm_test_paths_created {
my $args = shift;
my @e = map { File::Spec->catdir($args->{startdir}, File::Spec->splitdir($_)) }
($args->{test_dest}, $args->{test_base});
# destination # starting point
for my $e (@e) {
ok(-d $e, "Previously created $e for testing");
}
return @e;
}

sub confirm_chdir_via_expected_relpath {
my $args = shift;
chdir $args->{e}->[1] or die "Unable to chdir to $args->{e}->[1]: $!";
chdir $args->{expected_relpath} or die "Unable to chdir via relative path: $!";
my $thisdir = cwd();
my $expected_thisdir =
File::Spec->catdir($args->{startdir}, File::Spec->splitdir($args->{dest_real}));
is($thisdir, $expected_thisdir,
"chdir via relative path brought us to $expected_thisdir, as expected");
chdir $args->{startdir} or die "Unable to get back to $args->{startdir}: $!";
return $expected_thisdir;
}

sub calculate_relpath_via_abs2rel {
my $args = shift;
chdir $args->{e}->[1] or die "Unable to chdir to $args->{e}->[1]: $!";
my $calculated_relpath = File::Spec->abs2rel(@{$args->{e}});
is($calculated_relpath, $args->{expected_relpath},
"abs2rel() calculated relative path ($calculated_relpath) as expected ($args->{expected_relpath})");
chdir $calculated_relpath or die "Unable to chdir via calculated relative path: $!";
my $thisdir = cwd();
is($thisdir, $args->{expected_thisdir},
"chdir via calculated relative path ($calculated_relpath) brought us to $args->{expected_thisdir}, as expected");
chdir $args->{startdir} or die "Unable to get back to $args->{startdir}: $!";
return 1;
}
```

File: util/perl/OpenSSL/Test.pm
Comment (mspncp) [inline util/perl/OpenSSL/Test.pm]: Just one question for curiousity: why do you sometimes use `abs_path()`, and sometimes `rel2abs()`?

**Review by mspncp — APPROVED** [review]
I did some testing on OpenSSL_1_1_1-stable (partially because it compiles much faster, and partially because the unknown OpenSSL issue was probably encountered with the 1.1.1 release) comparing the difference of @levitte's fix vs. just reverting the abs2rel workaround:

* 7be676bd29 ([revert-abs2rel-workaround-111](https://github.com/mspncp/openssl/commits/revert-abs2rel-workaround-111)) Revert "Adding Test.pm with workaround for Perl abs2rel bug"
* c356838136 ([bring-back-relative-paths-111](https://github.com/mspncp/openssl/commits/bring-back-relative-paths-111)) OpenSSL::Test: bring back the relative paths

You find the branches on my clone [mspncp/openssl](https://github.com/mspncp/openssl):

```
* 7be676bd29 (revert-abs2rel-workaround-111) Revert "Adding Test.pm with workaround for Perl abs2rel bug"
| * c356838136 (HEAD -> bring-back-relative-paths-111) OpenSSL::Test: bring back the relative paths
|/  
* 8c7d505f87 (enable-debug-111) Enable debug
* f50f2725c0 (origin/OpenSSL_1_1_1-stable, github/OpenSSL_1_1_1-stable, OpenSSL_1_1_1-stable) Remove remaining references to crypto/include
...
```

_(**BTW:** did you notice that  GitHub automagically blends the commits from my clone into openssl/openssl, even though I didn't create a pull request for them?)_

The only difference I can see in the diffs is an unnecessary `../test` in the BLDTEST path. The relative pathes in the shlib_wrap.sh calls are identical.

Well, ok: Allthough I still don't see why the initial abs2rel workaround was necessary, together with this fix the change is acceptable for me. Not only does this pull request fix the unnecessary `../test` in the the `BLDTEST` path, but also makes the `SRCTOP` and `BLDTOP` debug output more explict:


```diff
--- build-revert-abs2rel-workaround-111/test.log        2020-01-24 23:40:38.665745348 +0100
+++ build-bring-back-relative-paths-111/test.log        2020-01-24 23:44:46.645959147 +0100
@@ -22,41 +22,41 @@
 run_tests: bldtop =../.
 ../../../openssl-1.1.1/test/recipes/01-test_abort.t .................... 
 DEBUG: __cwd(), directories and files:
-  $directories{BLDTEST} = "../../test"
+  $directories{BLDTEST} = ".."
   $directories{SRCTEST} = "../../../../openssl-1.1.1/test"
   $directories{SRCDATA} = "../../../../openssl-1.1.1/test/recipes/01-test_abort_data"
   $directories{RESULTS} = "."
   $directories{BLDAPPS} = "../../apps"
   $directories{SRCAPPS} = "../../../../openssl-1.1.1/apps"
-  $directories{SRCTOP}  = "../../../../openssl-1.1.1"
-  $directories{BLDTOP}  = "../.."
+  $directories{SRCTOP}  = "/home/msp/src/openssl-1.1.1"
+  $directories{BLDTOP}  = "/home/msp/src/external/build"
 
   current directory is "."
   the way back is ".."
 1..1
-DEBUG[__decorate_cmd]: $cmdstr = "../../util/shlib_wrap.sh ../../test/aborttest"
-DEBUG[__decorate_cmd]: $display_cmd = "../../util/shlib_wrap.sh ../../test/aborttest"
+DEBUG[__decorate_cmd]: $cmdstr = "../../util/shlib_wrap.sh ../aborttest"
+DEBUG[__decorate_cmd]: $display_cmd = "../../util/shlib_wrap.sh ../aborttest"
 ../../openssl-1.1.1/test/aborttest.c:14: OpenSSL internal error: Voluntary abort
-../../util/shlib_wrap.sh ../../test/aborttest => 134
+../../util/shlib_wrap.sh ../aborttest => 134
 ok 1 - Testing that abort is caught correctly
```


**Note:** The build logs were produced using the following script:


```bash
#!/bin/bash

set -o errexit

# note: current directory for all test runs is: /home/msp/src/external/build/test/test-runs

external=../external

rm -rf $external
mkdir -p $external

for branch in revert-abs2rel-workaround-111 bring-back-relative-paths-111; do
	git checkout $branch
	git rebase HEAD~ --onto enable-debug-111
done

for branch in enable-debug-111 revert-abs2rel-workaround-111 bring-back-relative-paths-111; do
	git checkout $branch
	mkdir -p $external/build
	(cd $external/build && ../../openssl-1.1.1/config && make && make V=1 test |& tee test.log)
	mv $external/build $external/build-$branch
done
```

If anyone is interested in the full logs, please let me know.

File: util/perl/OpenSSL/Test.pm
Comment (levitte) [reply util/perl/OpenSSL/Test.pm]: Regarding `abs_path`, I blame @mpeylo :wink:
We could absolutely use `rel2abs` instead

**Comment by levitte** [comment]
Merged.

master:
612539e8a678c6099131dfd0e5e4b85fa774eb1a OpenSSL::Test: bring back the relative paths

1.1.1:
c8943eb04ae41506ec8a1869103e2e1dec6bfcd8 OpenSSL::Test: bring back the relative paths

**Comment by mspncp** [comment]
Thank you for fixing it!

**Pull request closed without merging**
